src/patcher.c

```diff
@@ -563,25 +544,64 @@ init_patcher(void)
 }
 
 /*
- * create_movabs_r11
+ * create_movabs
  * Generates a movabs instruction, that assigns a 64 bit constant to
- * the R11 register.
+ * the 64 general purpose register.
+ * the reg_bits value must contain the X86 encoding of the register.
+ * Returns a pointer to the char right after the generated instruction.
  */
-static void
-create_movabs_r11(unsigned char *code, uint64_t value)
+static unsigned char *
+create_movabs(unsigned char *code, uint64_t value, unsigned char reg_bits)
 {
+	assert(reg_bits < 16);
+
 	unsigned char *bytes = (unsigned char *)&value;
 
-	code[0] = 0x49; /* movabs opcode */
-	code[1] = 0xbb; /* specifiy r11 as destination */
-	code[2] = bytes[0];
-	code[3] = bytes[1];
-	code[4] = bytes[2];
-	code[5] = bytes[3];
-	code[6] = bytes[4];
-	code[7] = bytes[5];
-	code[8] = bytes[6];
-	code[9] = bytes[7];
+	*code++ = 0x48 | (reg_bits >> 3); /* REX prefix */
+	*code++ = 0xb8 | (reg_bits & 7); /* opcode */
+	*code++ = bytes[0];
+	*code++ = bytes[1];
+	*code++ = bytes[2];
+	*code++ = bytes[3];
+	*code++ = bytes[4];
+	*code++ = bytes[5];
+	*code++ = bytes[6];
+	*code++ = bytes[7];
+
+	return code;
+}
+
+/*
+ * create_movabs_r11
+ * Generates a movabs instruction moving a value into the R11 register.
+ */
+static unsigned char *
+create_movabs_r11(unsigned char *code, uint64_t value)
+{
+	return create_movabs(code, value, 11);
+}
+
+/*
+ * relocate_instruction
+ * Places an instruction equivalent to `ins` to the memory location at `dst`.
+ * Only handles instructions that can be copied verbatim, and some LEA
+ * instructions.
+ */
+static unsigned char *
+relocate_instruction(unsigned char *dst,
+			const struct intercept_disasm_result *ins)
+{
+	if (ins->is_lea_rip) {
+		/*
+		 * Substitue a "lea $offset(%rip), %reg" instruction
+		 * by a movabs instruction, to achieve the same effect.
```

Comment:
Can you really do this? lea computes an address and stores it in the destination register, but mov also dereferences the computed address and stores the result in the destination register.

Reply:
Oh, right, I'll just change the comment to make it more clear.
e.g.:
`lea 1234(%rip), %rdi`, where `%rip` has the value `2222`
is changed to:
`movabs %rdi, 3456` (where 3456 = 1234 + 2222)
So we just compute the result of the lea instruction, and put it into the immediate value used in `movabs`. So it is a `movabs` with an immediate operand, doesn't dereference anything.